File: 2024/06/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

type position [2]int

type cursor struct {
	position
	vector [2]int
}

func (c *cursor) turn() {
	c.vector = [2]int{c.vector[1], -c.vector[0]}
}

func (c cursor) nextPos() [2]int {
	return [2]int{
		c.position[0] + c.vector[0],
		c.position[1] + c.vector[1]}
}

type region [][]bool

func (r region) isMapped(pos [2]int) bool {
	if (pos[0] >= 0 && pos[0] < len(r)) &&
		(pos[1] >= 0 && pos[1] < len(r[pos[0]])) {
		return true
	}

	return false
}

func (r region) solve() (visited map[[2]int]struct{}, isLoop bool) {
	curr := start
	visitedMap := map[[2]int]struct{}{}
	seenObstacles := map[[2]int]map[[2]int]struct{}{}

	for r.isMapped(curr.position) {
		for nextPos := curr.nextPos(); r.isMapped(nextPos) && r[nextPos[0]][nextPos[1]]; nextPos = curr.nextPos() {
			var seenVectors map[[2]int]struct{}

			if v, ok := seenObstacles[nextPos]; !ok {
				seenVectors = map[[2]int]struct{}{}
				seenObstacles[nextPos] = seenVectors
			} else {
				seenVectors = v
			}

			if _, ok := seenVectors[curr.vector]; ok {
				return visitedMap, true
			}

			seenVectors[curr.vector] = struct{}{}
			curr.turn()
		}

		visitedMap[curr.position] = struct{}{}
		curr.position = curr.nextPos()
	}

	return visitedMap, false
}

func (r region) solveWithObstacle(pos [2]int) (visited map[[2]int]struct{}, isLoop bool) {
	r[pos[0]][pos[1]] = true
	visited, isLoop = r.solve()
	r[pos[0]][pos[1]] = false
	return
}

var data = region([][]bool{})
var start = cursor{}

func parse(r io.Reader) {
	scanner := bufio.NewScanner(r)

	for i := 0; scanner.Scan(); i++ {
		row := []bool{}
		for j, r := range scanner.Text() {
			if r == '#' {
				row = append(row, true)
			} else {
				row = append(row, false)

				if r == '.' {
					continue
				}

				switch r {
				case '^':
					start.vector = [2]int{-1, 0}
				case '>':
					start.vector = [2]int{0, +1}
				case 'v':
					start.vector = [2]int{+1, 0}
				case '<':
					start.vector = [2]int{0, -1}
				}

				start.position = [2]int{i, j}
			}
		}

		data = append(data, row)
	}
}

func main() {
	parse(os.Stdin)

	visited, _ := data.solve()

	P1 := len(visited)
	
	P2 := 0
	for pos := range visited {
		if _, isLoop := data.solveWithObstacle(pos); isLoop {
			P2++
		}
	}

	fmt.Printf("P1: %d\n", P1)
	fmt.Printf("P2: %d\n", P2)
}

File: 2024/06/python/main.py
import fileinput
from collections.abc import Callable
from copy import copy
from typing import Literal, NamedTuple


class Vector(NamedTuple):
    c_row: Literal[-1, 0, +1]
    c_col: Literal[-1, 0, +1]


class Position(NamedTuple):
    row: int
    col: int


class Cursor:
    def __init__(self, pos: Position, vector: Vector):
        self.position: Position = pos
        self.vector = vector

    def turn(self):
        self.vector = Vector(self.vector.c_col, -self.vector.c_row)

    def next_pos(self) -> Position:
        row, col = self.position
        c_row, c_col = self.vector
        return Position(row + c_row, col + c_col)


class Region(list[list[bool]]):
    def guard_path(self, curr: Cursor) -> tuple[set[Position], bool]:
        curr = copy(curr)
        res: set[Position] = set()
        obstacles: dict[Position, set[Vector]] = {}

        def is_inside(pos: Position) -> bool:
            return 0 <= pos.row < len(self) and 0 <= pos.col < len(self[pos.row])

        while is_inside(curr.position):
            while is_inside(n_pos := curr.next_pos()) and self[n_pos.row][n_pos.col]:
                if curr.vector in (seen_vectors := obstacles.setdefault(n_pos, set())):
                    return res, True

                seen_vectors.add(curr.vector)
                curr.turn()

            res.add(curr.position)
            curr.position = n_pos

        return res, False

    def set_obstacle(self, pos: Position, func: Callable, *args, **kwargs):
        self[pos.row][pos.col] = True
        res = func(*args, **kwargs)
        self[pos.row][pos.col] = False
        return res


start: Cursor | None = None
data: Region = Region()

with fileinput.input() as file:
    for i, line in enumerate(file):
        parsed_line: list[bool] = []
        for j, c in enumerate(line[:-1]):
            if c == "#":
                parsed_line.append(True)
            else:
                parsed_line.append(False)

                if c == ".":
                    continue

                match c:
                    case "^":
                        vec = Vector(-1, 0)
                    case ">":
                        vec = Vector(0, +1)
                    case "v":
                        vec = Vector(+1, 0)
                    case "<":
                        vec = Vector(0, -1)
                    case _:
                        raise AssertionError(f"unparsable char: {c}")

                start = Cursor(Position(i, j), vec)

        data.append(parsed_line)

assert isinstance(start, Cursor)
guard_pos, _ = data.guard_path(start)

print(f"P1: {len(guard_pos)}")
print(
    f"P2: {[data.set_obstacle(pos, data.guard_path,start)[1] for pos in guard_pos].count(True)}"
)
